### PR TITLE
fix: support defaulter with not-existing workspaces

### PR DESCRIPTION
**Pull request by GGabriele**
The new defaulter interacts with Kong API to retrieve
entities schemas. Doing such query against a not existing
workspace returns an error, blocking the diff operation in decK.

This makes sure decK can run 'diff' regardless of the workspace
by simply not retrieving entities schemas from Kong when the
workspace doesn't exist yet.

Solves https://github.com/Kong/deck/issues/601#issuecomment-1057647420

### CODE REVIEW

File: file/builder.go [hunk 1]
```diff
@@ -807,7 +807,16 @@ func (b *stateBuilder) getPluginSchema(pluginName string) (map[string]interface{
 	if schema, ok := b.schemasCache[pluginName]; ok {
 		return schema, nil
 	}
-	schema, err := b.client.Plugins.GetFullSchema(b.ctx, &pluginName)
+
+	exists, err := utils.WorkspaceExists(b.ctx, b.client)
+	if err != nil {
+		return nil, fmt.Errorf("getPluginSchema: checking if workspace exists: %w", err)
```
Comment (hbagdi) [inline file/builder.go:813]:
```suggestion
		return nil, fmt.Errorf("ensure workspace exists: %w", err)
```

File: utils/defaulter.go [hunk 2]
```diff
@@ -225,7 +225,11 @@ func getKongDefaulterWithClient(ctx context.Context, opts DefaulterOpts) (*Defau
 // 3. schema defaults coming from Admin API (excluded Konnect)
 // 4. hardcoded defaults under utils/constants.go (Konnect-only)
 func GetDefaulter(ctx context.Context, opts DefaulterOpts) (*Defaulter, error) {
-	if opts.Client != nil && !opts.DisableDynamicDefaults {
+	exists, err := WorkspaceExists(ctx, opts.Client)
+	if err != nil {
+		return nil, fmt.Errorf("defaulter: checking if workspace exists: %w", err)
```
Comment (hbagdi) [inline utils/defaulter.go:230]: same

**Review by hbagdi — COMMENTED** [review]
Is it possible to add an e2e test for this patch?

File: file/builder.go (hunk 1 above)
Comment (GGabriele) [reply file/builder.go]: done

File: utils/defaulter.go (hunk 2 above)
Comment (GGabriele) [reply utils/defaulter.go]: done

**Comment by GGabriele** [comment]
> addressing comments

done: https://github.com/Kong/deck/pull/608/commits/452c490830dce24ddd058450779a2c8924838cab

File: tests/integration/diff_test.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	"testing"
+
+	"github.com/kong/deck/utils"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_Diff_Workspace(t *testing.T) {
+	tests := []struct {
+		name          string
+		kongFile      string
```
Comment (hbagdi) [inline tests/integration/diff_test.go:15]:
```suggestion
		stateFile      string
```

**Comment by hbagdi** [comment]
While this doesn't throw an error and gets the user past the creation, this is not enough.
Kong must be able to serve default outside the workspace context.
This change has the potential to surprise some users: depending on the existence of the workspace, decK diff may error out or may not error out.

Having said that, please merge this to get moving.

**Comment by GGabriele** [comment]
@hbagdi can you please formally approve this?

**Review by hbagdi — APPROVED** [review]

**Pull request merged**